### ISSUE TITLE
🐛 --paging always appears to be broken

### ISSUE THREAD

**Issue by salmankhilji** [dandavison/delta#497]
The `--paging always` option appears to be broken.  My understanding is that if `--paging` is set to `always`, then the output from `delta` will always be fed into the pager, regardless of whether or not the output needs paging to begin with.  i.e., _both_ for very short and long output from git.

## Tesing Pager Configuration

To demonstrate the issue, let me first test the pager configuration on my machine.  Executing the following command starts `less` , waiting for the me to press ESC (repeatedly, 3 times) to get back to the terminal:

```
echo -e 'a\nb\nc' | less
```

## Repro Steps

I will now create a `git` commit that has very short output—short enough that the output does not need a pager.

```
salman@byteevo-scm:~/Scratch$ mkdir delta-bugreport
salman@byteevo-scm:~/Scratch$ cd delta-bugreport
salman@byteevo-scm:~/Scratch/delta-bugreport$ echo -e 'a\nb\nc' >README
salman@byteevo-scm:~/Scratch/delta-bugreport$ git init .
Initialized empty Git repository in /home/salman/Scratch/delta-bugreport/.git/
salman@byteevo-scm:~/Scratch/delta-bugreport$ git add -A
salman@byteevo-scm:~/Scratch/delta-bugreport$ git commit -m "initial check-in"
[master (root-commit) 9817663] initial check-in
 1 file changed, 3 insertions(+)
 create mode 100644 README
salman@byteevo-scm:~/Scratch/delta-bugreport$ GIT_PAGER='delta -s --paging always' git show 9817663 >/dev/tty
commit 9817663a2f7f69c7df13a3d268769b3345835266
Author: Salman A. Khilji <salman.khilji@byteevo.com>
Date:   Sat Jan 9 04:29:18 2021 -0800

    initial check-in


added: README
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

│    │                                                         │ 1  │a
│    │                                                         │ 2  │b
│    │                                                         │ 3  │c
salman@byteevo-scm:~/Scratch/delta-bugreport$ 
```

We observe within the above output that `delta` did _not_ launch a pager because the output was very short despite the fact that the `--paging always' option was specified.

Now, repeat the above experiment with the file created with the following command instead:

```
seq 100 >README
```

This time, running the following command (with the right hash, of course!) will indeed launch `less`:

```
GIT_PAGER='delta -s --paging always' git show 00aa00bb >/dev/tty
```
Here is my `.gitconfig`:

```
[user]
	name = Salman A. Khilji
	email = salman.khilji@byteevo.com
[color]
	ui = auto
[diff]
	tool = meld
[log]
	decorate = no
[color "status"]
	branch = green
[credential]
	helper = /usr/share/doc/git/contrib/credential/libsecret/git-credential-libsecret
```

Notice that I do _not_ have `delta` configured globally in my `.gitconfig`.

```
$ delta --version
delta 0.5.0
```

**Comment by dandavison** [comment]
Hi @salmankhilji, thanks for the extremely helpful report. I initially didn't understand what's going on, but I think I found the explanation.

TL;DR https://git-scm.com/docs/git-config#Documentation/git-config.txt-corepager

```
When the LESS environment variable is unset, Git sets it to FRX (if LESS environment variable is set, Git does not change it at all).
```

Would you agree that that explains your observations?

<details>
 I don't fully understand what's going on yet, but let's figure it out.

First, what values do you have for the relevant environment variables? `env | grep -E -i '(less|pager)'`?

Let's do all experiments with `unset LESS PAGER BAT_PAGER DELTA_PAGER` unless we say otherwise.

Now, I don't quite understand what's going on yet: I note that if we pipe into delta, then `git show HEAD | delta --paging=always` works as expected. But if we do any of the following (on your short input) it does not:

```
git -c core.pager='delta --paging=always' show HEAD
GIT_PAGER='delta --paging always' git show
GIT_PAGER='delta --paging always' git show > /dev/tty
```
I don't quite understand this. I notice that the behavior is the same with `GIT_PAGER='bat --paging always'` (unsurprising since delta copied its pager-invoking code from bat.)

OK, I think  this might be a clue. I believe the following shows that the behavior you're observing is not to do with delta (but tell me if I'm getting confused!):

```
delta-bugreport(master) env | grep -E -i '(less|pager)'
delta-bugreport(master) echo $?
1
delta-bugreport(master) GIT_PAGER='less -R' git show   # Doesn't page
delta-bugreport(master) git show | less -R # Does page

delta-bugreport(master અ ) seq 100 >README
delta-bugreport(master અ ) GIT_PAGER='less -R' git diff  # Does page
```

Any ideas what's going on? The behavior seems consistent with `git` itself setting the env var `LESS=--quit-if-one-screen` without us asking it to.

OK...I think that's it. Here: https://git-scm.com/docs/git-config#Documentation/git-config.txt-corepager

```
When the LESS environment variable is unset, Git sets it to FRX (if LESS environment variable is set, Git does not change it at all).
```

Would you agree that that explains your observations?
</details>

**Comment by salmankhilji** [comment]
Agreed!  Thanks for your help.  This bug may now be marked as invalid.

**Comment by feep** [comment]
To clarify, briefly, for the confused who find themselves here:

0. This problem is caused by `git` being sneaky with `less`, not by `delta`
1. It can be resolved / worked around by setting ` $LESS` to something, anything that works with your config
2. ...and setting `pager = delta --paging=always` in `[core]`

**Comment by jaltekruse** [comment]
@feep I was trying to set this up, curious if you think there are pitfalls to doing this in the gitconfig to centralize this workaround? Seems to work okay so far, but just curious about your thoughts

 [core]
     pager = env LESS="" delta --paging=always

**Comment by feep** [comment]
> curious if you think there are pitfalls to doing this in the gitconfig to centralize this workaround?

@jaltekruse I don’t need to set `$LESS` in my git config but I see no reason that wouldn’t work.

This is my current `...git/config`:

```
# ...trimmed

[core]
  pager = delta

[delta]
  light = true
  navigate = true # use n and N to move between diff sections
```

`$PAGER` is undefined, so delta, fish and git are fighting it out for what pages it. Looks like it gets to `less` eventually:

```
❯ ps -Af | rg less
less --RAW-CONTROL-CHARS --quit-if-one-screen
```

I have no idea where those less flags come from. But if you don’t have them, you can set them in your `$PAGER`.

`$LESS` is set in my shell. From my comment three months ago, that seems to be the key to the workaround.

```
❯ echo $LESS
-i
```

Two things:

0. It works beautifully; pages diffs if they don’t fit on screen, no pager if they fit.
0. That was three months ago, I have no idea if I did anything else to get it to work.

Delta kicks ass, thanks devs and makers.
